loader/loader.go

```diff
@@ -0,0 +1,281 @@
+// the List operation fails, an error will be returned.
+func GetReleasePipelineRun(release *v1alpha1.Release, cli client.Client, ctx context.Context) (*v1beta1.PipelineRun, error) {
+	pipelineRuns := &v1beta1.PipelineRunList{}
+	opts := []client.ListOption{
+		client.Limit(1),
+		client.MatchingLabels{
+			tekton.ReleaseNameLabel:      release.Name,
+			tekton.ReleaseNamespaceLabel: release.Namespace,
+		},
+	}
+
+	err := cli.List(ctx, pipelineRuns, opts...)
+	if err == nil && len(pipelineRuns.Items) > 0 {
+		return &pipelineRuns.Items[0], nil
+	}
+
+	return nil, err
+}
+
+// GetReleasePlan returns the ReleasePlan referenced by the given Release. If the ReleasePlan is not found or
+// the Get operation fails, an error will be returned.
+func GetReleasePlan(release *v1alpha1.Release, cli client.Client, ctx context.Context) (*v1alpha1.ReleasePlan, error) {
+	return getObject(release.Spec.ReleasePlan, release.Namespace, cli, ctx, &v1alpha1.ReleasePlan{})
+}
+
+// GetReleaseStrategy returns the ReleaseStrategy referenced by the given ReleasePlanAdmission. If the ReleaseStrategy
+// is not found or the Get operation fails, an error will be returned.
+func GetReleaseStrategy(releasePlanAdmission *v1alpha1.ReleasePlanAdmission, cli client.Client, ctx context.Context) (*v1alpha1.ReleaseStrategy, error) {
+	return getObject(releasePlanAdmission.Spec.ReleaseStrategy, releasePlanAdmission.Namespace, cli, ctx, &v1alpha1.ReleaseStrategy{})
+}
+
+// GetSnapshot returns the Snapshot referenced by the Release being processed. If the Snapshot is not found or the Get
+// operation fails, an error is returned.
+func GetSnapshot(name, namespace string, cli client.Client, ctx context.Context) (*applicationapiv1alpha1.Snapshot, error) {
+	return getObject(name, namespace, cli, ctx, &applicationapiv1alpha1.Snapshot{})
+}
+
+// GetSnapshotEnvironmentBinding returns the SnapshotEnvironmentBinding associated with the given ReleasePlanAdmission.
+// That association is defined by both the Environment and Application matching between the ReleasePlanAdmission and
+// the SnapshotEnvironmentBinding. If the Get operation fails, an error will be returned.
+func GetSnapshotEnvironmentBinding(releasePlanAdmission *v1alpha1.ReleasePlanAdmission, cli client.Client, ctx context.Context) (*applicationapiv1alpha1.SnapshotEnvironmentBinding, error) {
+	bindingList := &applicationapiv1alpha1.SnapshotEnvironmentBindingList{}
+	opts := []client.ListOption{
+		client.InNamespace(releasePlanAdmission.Namespace),
+		client.MatchingFields{"spec.environment": releasePlanAdmission.Spec.Environment},
+	}
+
+	err := cli.List(ctx, bindingList, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, binding := range bindingList.Items {
+		if binding.Spec.Application == releasePlanAdmission.Spec.Application {
+			return &binding, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// Composite functions
```

Comment:
I don't think we should add these composite functions. Just in my last PR I had to split apart an existing composite function (getting the binding resources). I don't think they help us and I think they are just going overboard with trying to simplify the adapter code. The extra level of indirection doesn't help much in my opinion... Having 3-4 get methods in a row in the adapter is very easy to follow